src/types/mmapped.jl

```diff
@@ -0,0 +1,76 @@
+struct MmappedTIFF{T <: Colorant, N, O <: Unsigned, A <: AbstractMatrix{T}} <: AbstractTIFF{T, N}
+    """
+    2d slices in the file"
+    """
+    chunks::Vector{A}
+    """
+    The associated tags for each slice in this array
+    """
+    ifds::Vector{IFD{O}}
+    """
+    The 2d slice size
+    """
+    sz2::Dims{2}
+end
+
+function MmappedTIFF{T,N}(file::TiffFile{O}, ifds::Vector{IFD{O}}) where {T, N, O <: Unsigned}
+    N ∈ (2, 3) || error("only 2d and 3d TIFF arrays are supported")
+    if N == 2
+        length(ifds) == 1 || error("2d files must have a single Image File Directory")
+    end
+    ifd = first(ifds)
+    sz = (nrows(ifd), ncols(ifd))
+    compression = getdata(CompressionType, ifd, COMPRESSION, COMPRESSION_NONE)
+    compression == COMPRESSION_NONE || error("mmap is not yet supported with compression")
+    all(iscontiguous, ifds) || error("non-contiguous TIFFs are not yet supported by mmap")
+
+    seek(file, 0)
+    rawio = stream(file)
+    raw = Mmap.mmap(rawio, Vector{UInt8}, filesize(rawio))
+    chunks = [getchunk(T, raw, reverse(sz), ifd) for ifd in ifds]
+    return MmappedTIFF{T, N, O, eltype(chunks)}(chunks, ifds, sz)
```

Comment:
> I have access to a Windows machine, let me try to fix it locally first.

I believe https://github.com/timholy/TiffImages.jl/pull/1 fixes the test. But it seems to uncover a windows caveat that mmap-ed images should be surrounded by a `let` block and eager `GC.gc()`

https://github.com/timholy/TiffImages.jl/pull/2 is an attempt to eagerly do `GC.gc()` if issues like this happens. But one should still do `GC.gc()` before trying to remove the mmapped files.

Reply:
MWE:

```julia
using ImageCore
using TiffImages
using ImageShow

img = rand(Gray, 64, 64)
TiffImages.save("tmp.tif", img)

img = TiffImages.load("tmp.tif"; mmap=true)

img2 = TiffImages.load("tmp.tif") # works
TiffImages.save("tmp.tif", img2) # fails -- because img occupies the file handler
```

With https://github.com/timholy/TiffImages.jl/pull/2

```julia
img2 = let img = TiffImages.load("tmp.tif"; mmap=true)
    TiffImages.load("tmp.tif") # works
end
TiffImages.save("tmp.tif", img2) # works -- with an warning
```

```julia
┌ Warning: failed to open file tmp.tif in "w" mode, retry after GC.gc()
└ @ TiffImages d:\Julia\TiffImages.jl\src\utils.jl:205
```

Reply:
Maybe we should not use the `load` interface, but instead, use something like the following since mmap requests a resource so it should follow the RAII (Resource Acquisition Is Initialization) practice.

```julia
TiffImages.lazy_load(filepath) do img
    ...
end
```